src/core/statelisteners/searchlistener.js

```diff
@@ -0,0 +1,176 @@
+      callback: query => this._handleQueryUpdate(query)
+    });
+    return this;
+  }
+
+  /**
+   * Sets a search query and corresponding QUERY_TRIGGER on initialization. A prepopluated query
+   * in the URL takes precendence over a defaultInitialSearch.
+   */
+  _setQueryOnLoad () {
+    const prepopulatedQuery = this.storage.get(StorageKeys.QUERY);
+    if (prepopulatedQuery != null) {
+      if (this.storage.get(StorageKeys.QUERY_TRIGGER) !== QueryTriggers.SUGGEST) {
+        this.storage.setWithPersist(StorageKeys.QUERY_TRIGGER, QueryTriggers.QUERY_PARAMETER);
+      }
+    } else if (this.config.defaultInitialSearch != null) {
+      if (!this.config.defaultInitialSearch && !this.config.allowEmptySearch) {
+        return;
+      }
+      this.storage.setWithPersist(StorageKeys.QUERY_TRIGGER, QueryTriggers.INITIALIZE);
+      this.storage.set(StorageKeys.QUERY, this.config.defaultInitialSearch);
+    }
+  }
+
+  _handleQueryUpdate (query) {
+    if (query === null) {
+      if (this.config.defaultInitialSearch || this.config.defaultInitialSearch === '') {
+        this.storage.setWithPersist(StorageKeys.QUERY_TRIGGER, QueryTriggers.INITIALIZE);
+        this.storage.set(StorageKeys.QUERY, this.config.defaultInitialSearch);
+      }
+      return;
+    }
+    this._debouncedSearch(query);
+  }
+
+  /**
+   * Registers middleware, that is called before a search.
+   *
+   * @param {Function<Promise>} middlewareFunction
+   */
+  registerMiddleware (middlewareFunction) {
+    this.middleware.push(middlewareFunction);
+  }
+
+  /**
+   * Update the config of SearchListener. This function is needed for backwards compatibility for when
+   * this listener used to live in the SearchBar component.
+   *
+   * @param {Object} newConfig
+   */
+  updateConfig (newConfig) {
+    this.config = {
+      ...this.config,
+      ...newConfig
+    };
+  }
+
+  _debouncedSearch (query) {
+    if (this._throttled ||
+      (!query && !this.config.allowEmptySearch &&
+        this.storage.get(StorageKeys.QUERY_TRIGGER) !== QueryTriggers.FILTER_COMPONENT)) {
+      return;
+    }
+
+    this._throttled = true;
+    setTimeout(() => { this._throttled = false; }, this._searchCooldown);
+    return this._search(query);
+  }
+
+  _search (query) {
+    const middlewarePromises = this.middleware.map(middleware => middleware(query));
+    return Promise.all(middlewarePromises).then(() => {
+      if (this.config.verticalKey) {
+        return this.core.verticalSearch(this.config.verticalKey, this._getSearchOptions(), { input: query });
+      } else {
+        return this.core.search(query, this._getUrls(query), this._getSearchOptions());
+      }
+    });
+  }
+
+  /**
+   * Returns the current search options state, based on the current QUERY_TRIGGER.
+   *
+   * @returns {Object}
+   */
+  _getSearchOptions () {
+    const queryTrigger = this.storage.get(StorageKeys.QUERY_TRIGGER);
+    if (queryTrigger === QueryTriggers.FILTER_COMPONENT) {
```

Comment:
What do you think of having the pagination or filter component to be responsible for the search options when triggering a search? It would let us avoid having to update search listener when a component changes its search options

Reply:
I think that makes sense, but I also kind of like how doing it by query trigger categorizes the different search scenarios. Like, "when a filter makes a search, this is what happens," and you can just look at one place and get an idea of what happens for filters, instead of having to check every filter component to see what kind of thing it's doing. I guess this kind of encapsulates the logic? As opposed to having it in lots of different places. Plus we already have this notion of query trigger, so we don't have to add something new like  a QUERY_OPTIONS storage key (though I don't think this would be a big deal). 

Maybe this comes down to what kind of granularity we need for search options. Are we going to be adding new types of "search options" (and we should probably stop adding new ones in either case), and new features that we'll have to add one new query trigger each for? I think... probably not? Even then I... kind of like how it makes it easier to reuse the same logic, on the off chance we need to trigger that search in two different places it's a little easier to write. I could have just been looking at this for too long though

Reply:
I'm definitely a fan of how clean the interface is for components triggering a search. Where I'm coming from is that it maybe this mapping of keys to shouldn't be the responsibility of the search listener class? Maybe we have a better separation of concerns if we were to create a new class ( or a helper function ) whose sole responsibility is to determine the search options based on the QueryTrigger. 

However, it doesn't seem terribly out of place here so I don't mind leaving it here at all.

Reply:
Yeah I think we can pull it out to another class!

Reply:
actually I think there's a number of things that can be split out from this class